From 12fb9219923f41528f673d4f31667db2c3f103ff Mon Sep 17 00:00:00 2001
From: Richard Lowe <richlowe@richlowe.net>
Date: Sat, 23 Sep 2023 23:05:02 +0000
Subject: [PATCH] 16225 cputrack should not be isaexec'd Reviewed by: Andy
 Fiddaman <illumos@fiddaman.net> Reviewed by: Toomas Soome <tsoome@me.com>
 Approved by: Robert Mustacchi <rm@fingolfin.org>

---
 usr/src/cmd/cpc/common/cputrack.c             | 146 +-----------------
 usr/src/cmd/cpc/cputrack/Makefile             |  64 ++++----
 usr/src/cmd/cpc/cputrack/Makefile.com         |  65 --------
 usr/src/cmd/cpc/cputrack/amd64/Makefile       |  33 ----
 usr/src/cmd/cpc/cputrack/i386/Makefile        |  31 ----
 usr/src/cmd/cpc/cputrack/sparcv9/Makefile     |  33 ----
 .../pkg/manifests/diagnostic-cpu-counters.p5m |   7 +-
 7 files changed, 42 insertions(+), 337 deletions(-)
 delete mode 100644 usr/src/cmd/cpc/cputrack/Makefile.com
 delete mode 100644 usr/src/cmd/cpc/cputrack/amd64/Makefile
 delete mode 100644 usr/src/cmd/cpc/cputrack/i386/Makefile
 delete mode 100644 usr/src/cmd/cpc/cputrack/sparcv9/Makefile

diff --git a/usr/src/cmd/cpc/common/cputrack.c b/usr/src/cmd/cpc/common/cputrack.c
index 198cdd1a01..6b70fa5caa 100644
--- a/usr/src/cmd/cpc/common/cputrack.c
+++ b/usr/src/cmd/cpc/common/cputrack.c
@@ -87,10 +87,6 @@ cputrack_pctx_errfn(const char *fn, const char *fmt, va_list ap)
 static int cputrack(int argc, char *argv[], int optind);
 static void intr(int);
 
-#if defined(__i386)
-static void p4_ht_error(void);
-#endif
-
 #if !defined(TEXT_DOMAIN)
 #define	TEXT_DOMAIN	"SYS_TEST"
 #endif
@@ -366,17 +362,13 @@ pinit_lwp(pctx_t *pctx, pid_t pid, id_t lwpid, void *arg)
 	if (cpc_bind_pctx(cpc, pctx, lwpid, set, 0) != 0 ||
 	    cpc_set_sample(cpc, set, *data2) != 0) {
 		errstr = strerror(errno);
-		if (errno == EAGAIN)
+		if (errno == EAGAIN) {
 			(void) cpc_unbind(cpc, set);
-#if defined(__i386)
-		if (errno == EACCES)
-			p4_ht_error();
-		else
-#endif
-			(void) fprintf(stderr, gettext(
-			    "%6d: init_lwp: can't bind perf counters "
-			    "to lwp%d - %s\n"), (int)pid, (int)lwpid,
-			    errstr);
+		}
+
+		(void) fprintf(stderr, gettext(
+		    "%6d: init_lwp: can't bind perf counters "
+		    "to lwp%d - %s\n"), (int)pid, (int)lwpid, errstr);
 		return (-1);
 	}
 
@@ -750,132 +742,6 @@ cputrack(int argc, char *argv[], int optind)
 	return (err != 0 ? 1 : 0);
 }
 
-#if defined(__i386)
-
-#define	OFFLINE_CMD	"/usr/sbin/psradm -f "
-#define	BUFSIZE		5	/* enough for "n " where n is a cpuid */
-
-/*
- * cpc_bind_pctx() failed with EACCES, which means the user must first offline
- * all but one logical processor on each physical processor. Print to stderr the
- * psradm command string to do this.
- */
-static void
-p4_ht_error(void)
-{
-	kstat_ctl_t	*kc;
-	kstat_t		*ksp;
-	kstat_named_t	*k;
-	int		i;
-	int		max;
-	int		stat;
-	int		*designees;
-	int		*must_offline;
-	char		buf[BUFSIZE];
-	char		*cmd;
-	int		noffline = 0;
-	int		ndone = 0;
-
-	(void) fprintf(stderr, "%s\n",
-	    gettext("Pentium 4 processors with HyperThreading present.\nOffline"
-	    " all but one logical processor on each physical processor in"
-	    " order to use\ncputrack.\n"));
-
-
-	if ((kc = kstat_open()) == NULL)
-		return;
-
-	max = sysconf(_SC_CPUID_MAX);
-	if ((designees = malloc(max * sizeof (*designees))) == NULL) {
-		(void) fprintf(stderr, gettext("%s: no memory available\n"),
-		    opts->pgmname);
-		exit(0);
-	}
-
-	if ((must_offline = malloc(max * sizeof (*designees))) == NULL) {
-		(void) fprintf(stderr, gettext("%s: no memory available\n"),
-		    opts->pgmname);
-		exit(0);
-	}
-
-	for (i = 0; i < max; i++) {
-		designees[i] = -1;
-		must_offline[i] = 0;
-	}
-
-	for (i = 0; i < max; i++) {
-		stat = p_online(i, P_STATUS);
-		if (stat != P_ONLINE && stat != P_NOINTR)
-			continue;
-
-		if ((ksp = kstat_lookup(kc, "cpu_info", i, NULL)) == NULL) {
-			free(designees);
-			free(must_offline);
-			return;
-		}
-
-		if (kstat_read(kc, ksp, NULL) == -1) {
-			free(designees);
-			free(must_offline);
-			return;
-		}
-
-		if ((k = (kstat_named_t *)kstat_data_lookup(ksp, "chip_id"))
-		    == NULL) {
-			free(designees);
-			free(must_offline);
-			return;
-		}
-
-		if (designees[k->value.i32] == -1)
-			/*
-			 * This chip doesn't yet have a CPU designated to remain
-			 * online; let this one be it.
-			 */
-			designees[k->value.i32] = i;
-		else {
-			/*
-			 * This chip already has a designated CPU; this CPU must
-			 * go offline.
-			 */
-			must_offline[i] = 1;
-			noffline++;
-		}
-	}
-
-	/*
-	 * Now construct a string containing the command line used to offline
-	 * the appropriate processors.
-	 */
-
-	if ((cmd = malloc(strlen(OFFLINE_CMD) + (noffline * BUFSIZE) + 1))
-	    == NULL) {
-		(void) fprintf(stderr, gettext("%s: no memory available\n"),
-		    opts->pgmname);
-		exit(0);
-	}
-
-	(void) strcpy(cmd, OFFLINE_CMD);
-
-	for (i = 0; i < max; i++) {
-		if (must_offline[i] == 0)
-			continue;
-
-		ndone++;
-		(void) snprintf(buf, BUFSIZE, "%d", i);
-		if (ndone < noffline)
-			(void) strcat(buf, " ");
-		(void) strcat(cmd, buf);
-	}
-
-	(void) fprintf(stderr, "%s:\n%s\n", gettext("The following command "
-	    "will configure the system appropriately"), cmd);
-
-	exit(1);
-}
-
-#endif /* defined(__i386) */
-
 /*ARGSUSED*/
 static void
 intr(int sig)
diff --git a/usr/src/cmd/cpc/cputrack/Makefile b/usr/src/cmd/cpc/cputrack/Makefile
index 13989f85da..de0f441119 100644
--- a/usr/src/cmd/cpc/cputrack/Makefile
+++ b/usr/src/cmd/cpc/cputrack/Makefile
@@ -2,9 +2,8 @@
 # CDDL HEADER START
 #
 # The contents of this file are subject to the terms of the
-# Common Development and Distribution License, Version 1.0 only
-# (the "License").  You may not use this file except in compliance
-# with the License.
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
 #
 # You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 # or http://www.opensolaris.org/os/licensing.
@@ -19,50 +18,55 @@
 #
 # CDDL HEADER END
 #
+
 #
-# Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
+# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
 #
-# ident	"%Z%%M%	%I%	%E% SMI"
+# Copyright (c) 2018, Joyent, Inc.
 #
 
-PROG =	cputrack
+include	../../Makefile.cmd
+include	../../Makefile.cmd.64
+include	../../Makefile.ctf
 
-include ../../Makefile.cmd
+PROG =		cputrack
+OBJS =		$(PROG).o caps.o time.o setgrp.o strtoset.o
+LDLIBS +=	-lcpc -lpctx
 
-$(64ONLY)SUBDIRS= $(MACH)
-$(BUILD64)SUBDIRS += $(MACH64)
+CFLAGS +=	$(CCVERBOSE)
+CPPFLAGS +=	-I$(SRC)/lib/libcpc/common
 
-all	:=	TARGET = all
-install	:=	TARGET = install
-clean	:=	TARGET = clean
-clobber	:=	TARGET = clobber
-lint	:=	TARGET = lint
-strip	:=	TARGET = strip
+POFILES =	../common/$(PROG).po ../common/caps.po
+POFILE =	$(PROG)_cmd.po
+
+# not linted
+SMATCH=off
+
+ROOTLINK = $(PROG:%=$(ROOTBIN64)/%)
 
 .KEEP_STATE:
 
-all:	$(SUBDIRS)
+all: $(PROG)
 
-clean clobber lint:	$(SUBDIRS)
+install: all $(ROOTPROG) $(ROOTLINK)
 
-strip: all $(SUBDIRS)
+$(PROG): $(OBJS)
+	$(LINK.c) $(OBJS) -o $@ $(LDLIBS)
+	$(POST_PROCESS)
 
-CPPFLAGS +=	-I$(SRC)/lib/libcpc/common
-POFILES =	../common/$(PROG).po ../common/caps.po
-POFILE =	$(PROG)_cmd.po
+$(ROOTLINK):
+	$(RM) $@; $(SYMLINK) ../../bin/cputrack $@
+
+clean:
+	$(RM) $(OBJS)
+
+%.o:	../common/%.c
+	$(COMPILE.c) $<
+	$(CTFCONVERT_O)
 
 $(POFILE): $(POFILES)
 	$(RM) $@
 	cat $(POFILES) > $@
 
-install:	$(SUBDIRS)
-	-$(RM) $(ROOTPROG)
-	-$(LN) $(ISAEXEC) $(ROOTPROG)
-
-$(SUBDIRS):	FRC
-	@cd $@; pwd; $(MAKE) $(TARGET)
-
-FRC:
-
 include ../../Makefile.targ
diff --git a/usr/src/cmd/cpc/cputrack/Makefile.com b/usr/src/cmd/cpc/cputrack/Makefile.com
deleted file mode 100644
index 8226e1a57c..0000000000
--- a/usr/src/cmd/cpc/cputrack/Makefile.com
+++ /dev/null
@@ -1,65 +0,0 @@
-#
-# CDDL HEADER START
-#
-# The contents of this file are subject to the terms of the
-# Common Development and Distribution License (the "License").
-# You may not use this file except in compliance with the License.
-#
-# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
-# or http://www.opensolaris.org/os/licensing.
-# See the License for the specific language governing permissions
-# and limitations under the License.
-#
-# When distributing Covered Code, include this CDDL HEADER in each
-# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
-# If applicable, add the following below this CDDL HEADER, with the
-# fields enclosed by brackets "[]" replaced with your own identifying
-# information: Portions Copyright [yyyy] [name of copyright owner]
-#
-# CDDL HEADER END
-#
-#
-# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
-# Use is subject to license terms.
-#
-# Copyright (c) 2018, Joyent, Inc.
-
-include	../../../Makefile.cmd
-
-PROG =		cputrack
-OBJS =		$(PROG).o caps.o time.o setgrp.o strtoset.o
-SRCS =		$(OBJS:%.o=../../common/%.c)
-LDLIBS +=	-lcpc -lpctx
-
-CFLAGS +=	$(CCVERBOSE) $(CTF_FLAGS)
-CFLAGS64 +=	$(CCVERBOSE) $(CTF_FLAGS)
-CPPFLAGS +=	-I$(SRC)/lib/libcpc/common
-
-# not linted
-SMATCH=off
-
-LINTFLAGS +=	-u
-LINTFLAGS64 +=	-u
-
-.KEEP_STATE:
-
-all: $(PROG)
-
-$(PROG): $(OBJS)
-	$(LINK.c) $(OBJS) -o $@ $(LDLIBS)
-	$(POST_PROCESS)
-	$(CTFMERGE) -L VERSION -o $@ $(OBJS)
-
-clean:
-	$(RM) $(OBJS)
-
-lint:	lint_SRCS
-
-strip:
-	$(STRIP) $(PROG)
-
-%.o:	../../common/%.c
-	$(COMPILE.c) $<
-	$(CTFCONVERT_O)
-
-include	../../../Makefile.targ
diff --git a/usr/src/cmd/cpc/cputrack/amd64/Makefile b/usr/src/cmd/cpc/cputrack/amd64/Makefile
deleted file mode 100644
index 3794e14c2c..0000000000
--- a/usr/src/cmd/cpc/cputrack/amd64/Makefile
+++ /dev/null
@@ -1,33 +0,0 @@
-#
-# CDDL HEADER START
-#
-# The contents of this file are subject to the terms of the
-# Common Development and Distribution License, Version 1.0 only
-# (the "License").  You may not use this file except in compliance
-# with the License.
-#
-# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
-# or http://www.opensolaris.org/os/licensing.
-# See the License for the specific language governing permissions
-# and limitations under the License.
-#
-# When distributing Covered Code, include this CDDL HEADER in each
-# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
-# If applicable, add the following below this CDDL HEADER, with the
-# fields enclosed by brackets "[]" replaced with your own identifying
-# information: Portions Copyright [yyyy] [name of copyright owner]
-#
-# CDDL HEADER END
-#
-#
-# Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
-# Use is subject to license terms.
-#
-# ident	"%Z%%M%	%I%	%E% SMI"
-#
-
-include ../Makefile.com
-
-include ../../../Makefile.cmd.64
-
-install: all $(ROOTPROG64)
diff --git a/usr/src/cmd/cpc/cputrack/i386/Makefile b/usr/src/cmd/cpc/cputrack/i386/Makefile
deleted file mode 100644
index 139c8bd92a..0000000000
--- a/usr/src/cmd/cpc/cputrack/i386/Makefile
+++ /dev/null
@@ -1,31 +0,0 @@
-#
-# CDDL HEADER START
-#
-# The contents of this file are subject to the terms of the
-# Common Development and Distribution License (the "License").
-# You may not use this file except in compliance with the License.
-#
-# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
-# or http://www.opensolaris.org/os/licensing.
-# See the License for the specific language governing permissions
-# and limitations under the License.
-#
-# When distributing Covered Code, include this CDDL HEADER in each
-# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
-# If applicable, add the following below this CDDL HEADER, with the
-# fields enclosed by brackets "[]" replaced with your own identifying
-# information: Portions Copyright [yyyy] [name of copyright owner]
-#
-# CDDL HEADER END
-#
-#
-# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
-# Use is subject to license terms.
-#
-#ident	"%Z%%M%	%I%	%E% SMI"
-
-include ../Makefile.com
-
-LDLIBS +=	-lkstat
-
-install: all $(ROOTPROG32)
diff --git a/usr/src/cmd/cpc/cputrack/sparcv9/Makefile b/usr/src/cmd/cpc/cputrack/sparcv9/Makefile
deleted file mode 100644
index 3794e14c2c..0000000000
--- a/usr/src/cmd/cpc/cputrack/sparcv9/Makefile
+++ /dev/null
@@ -1,33 +0,0 @@
-#
-# CDDL HEADER START
-#
-# The contents of this file are subject to the terms of the
-# Common Development and Distribution License, Version 1.0 only
-# (the "License").  You may not use this file except in compliance
-# with the License.
-#
-# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
-# or http://www.opensolaris.org/os/licensing.
-# See the License for the specific language governing permissions
-# and limitations under the License.
-#
-# When distributing Covered Code, include this CDDL HEADER in each
-# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
-# If applicable, add the following below this CDDL HEADER, with the
-# fields enclosed by brackets "[]" replaced with your own identifying
-# information: Portions Copyright [yyyy] [name of copyright owner]
-#
-# CDDL HEADER END
-#
-#
-# Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
-# Use is subject to license terms.
-#
-# ident	"%Z%%M%	%I%	%E% SMI"
-#
-
-include ../Makefile.com
-
-include ../../../Makefile.cmd.64
-
-install: all $(ROOTPROG64)
diff --git a/usr/src/pkg/manifests/diagnostic-cpu-counters.p5m b/usr/src/pkg/manifests/diagnostic-cpu-counters.p5m
index bbc5b09bea..9139d8156e 100644
--- a/usr/src/pkg/manifests/diagnostic-cpu-counters.p5m
+++ b/usr/src/pkg/manifests/diagnostic-cpu-counters.p5m
@@ -33,11 +33,8 @@ set name=info.classification value=org.opensolaris.category.2008:System/Core
 set name=variant.arch value=$(ARCH)
 dir  path=usr group=sys
 dir  path=usr/bin
-$(i386_ONLY)dir path=usr/bin/$(ARCH32)
-$(i386_ONLY)file path=usr/bin/$(ARCH32)/cputrack mode=0555
-dir  path=usr/bin/$(ARCH64)
-file path=usr/bin/$(ARCH64)/cputrack mode=0555
-hardlink path=usr/bin/cputrack target=../lib/isaexec
+link path=usr/bin/$(ARCH64)/cputrack target=../../bin/cputrack
+file path=usr/bin/cputrack mode=0555
 dir  path=usr/include
 file path=usr/include/libcpc.h
 file path=usr/include/libpctx.h
-- 
2.46.0

